src/iosource/Packet.cc

```diff
@@ -277,12 +277,38 @@ void Packet::ProcessLayer2()
 		pdata += rtheader_len;
 
 		int type_80211 = pdata[0];
-		int len_80211 = 0;
+		u_char len_80211 = 0;
 		if ( (type_80211 >> 4) & 0x04 )
 			{
 			//identified a null frame (we ignore for now).  no weird.
 			return;
 			}
+
+		// Look for data frames
+		if ( type_80211 & 0x08 )
+			{
+			// Determine link-layer addresses based
+			// on 'To DS' and 'From DS' flags
+			switch ( pdata[1] & 0x03 ) {
+				case 0x00:
+					l2_dst = pdata + 4;
+					l2_src = pdata + 10;
+					break;
+				case 0x01:
+					l2_dst = pdata + 16;
+					l2_src = pdata + 10;
+					break;
+				case 0x02:
+					l2_dst = pdata + 4;
+					l2_src = pdata + 16;
+					break;
+				case 0x03:
+					l2_dst = pdata + 16;
+					l2_src = pdata + 24;
```

Comment:
Somehow both: 802.11 defines 3 frame types (management, control, data) and multiple subtypes for each frame type. Depending on the type and subtype, a frame header might or might not contain certain fields. For the address fields also the semantics vary (that's what this code deals with). To me the given code seems to assume that Bro will only see data frames. For data frames it handles the cases with and without QoS information.